WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,443 @@
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public record DependencyGraph
+	{
+		public const int K = ProtocolConstants.CredentialNumber;
+
+		public static ImmutableArray<CredentialType> CredentialTypes { get; } = ImmutableArray.Create<CredentialType>(CredentialType.Amount, CredentialType.Vsize);
+
+		public ImmutableList<RequestNode> Vertices { get; private set; } = ImmutableList<RequestNode>.Empty;
+
+		public ImmutableList<RequestNode> Inputs { get; private set; } = ImmutableList<RequestNode>.Empty;
+
+		public ImmutableList<RequestNode> Outputs { get; private set; } = ImmutableList<RequestNode>.Empty;
+
+		public ImmutableList<RequestNode> Reissuances { get; private set; } = ImmutableList<RequestNode>.Empty; // TODO sort
+
+		// Internal properties used to keep track of effective values and edges
+		public ImmutableSortedDictionary<CredentialType, CredentialEdgeSet> edgeSets { get; init; } = ImmutableSortedDictionary<CredentialType, CredentialEdgeSet>.Empty
+			.Add(CredentialType.Amount, new() { CredentialType = CredentialType.Amount })
+			.Add(CredentialType.Vsize, new() { CredentialType = CredentialType.Vsize });
+
+		public long Balance(RequestNode node, CredentialType credentialType) => edgeSets[credentialType].Balance(node);
+
+		public IEnumerable<CredentialDependency> InEdges(RequestNode node, CredentialType credentialType) => edgeSets[credentialType].InEdges(node);
+
+		public IEnumerable<CredentialDependency> OutEdges(RequestNode node, CredentialType credentialType) => edgeSets[credentialType].OutEdges(node);
+
+		public int InDegree(RequestNode node, CredentialType credentialType) => edgeSets[credentialType].InDegree(node);
+
+		public int OutDegree(RequestNode node, CredentialType credentialType) => edgeSets[credentialType].OutDegree(node);
+
+		/// <summary>Construct a graph from amounts, and resolve the
+		/// credential dependencies.</summary>
+		///
+		/// <remarks>Should only produce valid graphs. The elements of the
+		/// <see>Vertices</see> property will correspond to the given values in order,
+		/// and may contain additional nodes if reissuance requests are
+		/// required.</remarks>
+		public static DependencyGraph ResolveCredentialDependencies(IEnumerable<IEnumerable<ulong>> inputValues, IEnumerable<IEnumerable<ulong>> outputValues)
+		{
+			return FromValues(inputValues, outputValues).ResolveCredentials();
+		}
+
+		public static DependencyGraph FromValues(IEnumerable<IEnumerable<ulong>> inputValues, IEnumerable<IEnumerable<ulong>> outputValues)
+		{
+			if (Enumerable.Concat(inputValues, outputValues).Any(x => x.Count() != CredentialTypes.Length))
+			{
+				throw new ArgumentException($"Number of credential values must be {CredentialTypes.Length}");
+			}
+
+			foreach (var credentialType in CredentialTypes)
+			{
+				// no Sum(Func<ulong, ulong>)) variant
+				long credentialValue(IEnumerable<ulong> x) => (long)x.Skip((int)credentialType).First();
```

Comment:
I would rename the function and also use the `ElementAt` instead of skip+first. Also, `x` is not a good name for an argument even when it is the first choice for lambda expressions.

```suggestion
				long credentialTypeValue(IEnumerable<ulong> x) => (long)x.ElementAt((int)credentialType)
```

It is also possible to do:
```c#
				long credentialTypeValue(IEnumerable<ulong> x) => x.Cast<long>().ElementAt((int)credentialType)
```